storage/innobase/lock/lock0lock.cc

```diff
@@ -6428,3 +6443,86 @@ void lock_update_split_and_merge(
                           PAGE_HEAP_NO_SUPREMUM,
                           lock_get_min_heap_no(right_block));
 }
+#ifdef WITH_WSREP
+/*
+  Wsrep BF watchdog implementation:
+
+  We use InnoDB monitor thread to call wsrep_run_BF_lock_wait_watchdog()
+  periodically to check if there are applier threads hanging in
+  lock wait. If the lock wait has taken more than 5 seconds, the
+  potential local blocking transactions are killed until the
+  applier thread is released.
+ */
+
+std::atomic<uint> wsrep_BF_waiting_count;
+uint innodb_wsrep_applier_lock_wait_timeout =
+    INNODB_WSREP_APPLIER_LOCK_WAIT_TIMEOUT_DEFAULT;
+
+struct wsrep_BF_lock_waiters_t
+{
+  std::vector<trx_t *, ut_allocator<trx_t *> > bf_waiters{};
+  ulonglong max_wait_time{0};
+  my_hrtime_t now{my_hrtime_coarse()};
+};
+
+static my_bool wsrep_BF_lock_waiters_collect(rw_trx_hash_element_t *element,
+                                             wsrep_BF_lock_waiters_t *waiters)
+{
+  element->mutex.wr_lock();
+  auto *trx= element->trx;
+  if (wsrep_thd_is_BF(trx->mysql_thd, false) && trx->lock.wait_lock)
+  {
+    const auto timeout= innodb_wsrep_applier_lock_wait_timeout;
+    const auto wait_time=
+      (waiters->now.val - trx->lock.suspend_time.load().val)/HRTIME_RESOLUTION;
+    if (wait_time >= timeout)
+    {
+      try
+      {
+        waiters->bf_waiters.push_back(trx);
+        /* Grab reference to prevent trx going out of scope before
+           it is explicitly released in wsrep_run_BF_lock_wait_watchdog(). */
+        trx->reference();
+      }
+      catch (const std::bad_alloc &)
+      {
+        ib::warn() << "WSREP: Failed to alloc memory for bf_waiters";
+        element->mutex.wr_unlock();
+        return 1;
+      }
+    }
+    waiters->max_wait_time= std::max(waiters->max_wait_time, wait_time);
+  }
+
+  element->mutex.wr_unlock();
+  return 0;
+}
+
+void wsrep_run_BF_lock_wait_watchdog()
+{
+  if (!wsrep_BF_waiting_count.load(std::memory_order_relaxed))
+  {
+    /* No BF threads waiting. */
+    return;
+  }
+
+  my_hrtime_t start_time= my_hrtime_coarse();
+
+  mysql_mutex_lock(&lock_sys.wait_mutex);
+  wsrep_BF_lock_waiters_t waiters;
+  trx_sys.rw_trx_hash.iterate_no_dups(wsrep_BF_lock_waiters_collect, &waiters);
+  mysql_mutex_unlock(&lock_sys.wait_mutex);
+
+  for (auto *trx : waiters.bf_waiters)
+  {
+    lock_wait_wsrep(trx);
+    trx->release_reference();
```

Comment:
Can `trx` be committed or rolled back meanwhile? What would happen in that case?